### PR TITLE
Avoid collisions in cs files generated by Grpc.Tools

### PR DESCRIPTION
**Pull request by jtattermusch**
Just an experiment for now.

### CODE REVIEW

**Comment by jtattermusch** [comment]
Adhoc artifact build: https://sponge.corp.google.com/invocation?id=85d55e0a-0188-4d42-92c3-d4cacb8ecc20&searchFor=

**Comment by jtattermusch** [comment]
CC @Falco20019  this is my first experiment to fix https://github.com/grpc/grpc/issues/17672 (haven't tested it yet).

**Comment by jtattermusch** [comment]
Current version gives `obj/Debug/net45/da39a3ee5e6b4b0d/ : error : No such file or directory [/var/local/git/grpc/test/distrib/csharp/DistribTest/DistribTestDotNet.csproj]` because `--csharp_out` and `--grpc_out` don't create the subdirectories automatically.

File: src/csharp/Grpc.Tools/GeneratorServices.cs [hunk 1]
```diff
@@ -69,14 +69,14 @@ public override string[] GetPossibleOutputs(ITaskItem protoItem)
...
 
             if (doGrpc)
             {
-                // Override outdir if kGrpcOutputDir present, default to proto output.
-                string grpcdir = protoItem.GetMetadata(Metadata.GrpcOutputDir);
+                // Override outdir if GrpcOutputDir present, default to proto output.
+                string grpcdir = DepFileUtil.GetOutputDirWithHash(protoItem.GetMetadata(Metadata.GrpcOutputDir), protoItem.ItemSpec);
                 filename = LowerUnderscoreToUpperCamelGrpcWay(basename);
                 outputs[1] = Path.Combine(
                     grpcdir != "" ? grpcdir : outdir, filename) + "Grpc.cs";
```
Comment (Falco20019) [inline src/csharp/Grpc.Tools/GeneratorServices.cs:82]: This won't be string.Empty anymore as it always includes the hash.

Comment (jtattermusch) [reply src/csharp/Grpc.Tools/GeneratorServices.cs:82]: Good catch.

**Comment by jtattermusch** [comment]
Superseded by #22869

**Pull request closed without merging**